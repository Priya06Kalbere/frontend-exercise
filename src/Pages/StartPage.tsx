import React, { useState } from 'react'
import { useNavigate } from 'react-router-dom'
import { BoardSize, Difficulty } from '../types'
import backgroundImage from '../../public/assets/LoginBackground.png'

interface StartPageProps {
  onStartGame: (playerName: string, boardSize: BoardSize, difficulty: Difficulty) => void
}

const boardSizes: BoardSize[] = [
  { rows: 2, cols: 2, name: '2x2' },
  { rows: 3, cols: 4, name: '3x4' },
  { rows: 4, cols: 4, name: '4x4' },
  { rows: 4, cols: 5, name: '4x5' }
]

const allDifficulties: Difficulty[] = ['easy', 'medium', 'hard']

const difficultyDesc: Record<Difficulty, string> = {
  easy: 'Beginner-friendly',
  medium: 'Balanced challenge',
  hard: 'High difficulty'
}

// Function to calculate move limit based on board size and difficulty
const getMoveLimit = (difficulty: Difficulty, boardSize: BoardSize) => {
  const totalPairs = (boardSize.rows * boardSize.cols) / 2

  let base: number
  if (boardSize.rows * boardSize.cols <= 8) {
    base = { easy: 5, medium: 4, hard: 3 }[difficulty]
  } else if (boardSize.rows * boardSize.cols <= 16) {
    base = { easy: 6, medium: 5, hard: 4 }[difficulty]
  } else {
    base = { easy: 7, medium: 6, hard: 5 }[difficulty]
  }

  return Math.floor(base * totalPairs)
}

const StartPage: React.FC<StartPageProps> = ({ onStartGame }) => {
  const [inputName, setInputName] = useState('')
  const [selectedSize, setSelectedSize] = useState<BoardSize>(boardSizes[1])
  const [difficulty, setDifficulty] = useState<Difficulty>('easy')
  const [nameError, setNameError] = useState(false)
  const navigate = useNavigate()

  const handleNameChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const value = e.target.value
    setInputName(value)
    if (nameError && value.trim().length > 0) {
      setNameError(false)
    }
  }

  const handleStart = () => {
    if (!inputName.trim()) {
      setNameError(true)
      return
    }
    setNameError(false)
    onStartGame(inputName.trim(), selectedSize, difficulty)
    navigate('/game')
  }

  return (
    <div className='flex h-screen items-center justify-center bg-cover' style={{ backgroundImage: `url(${backgroundImage})` }}>
      <div
        className='w-full max-w-md rounded-xl border border-pink-400/50 px-8 py-10 shadow-2xl'
        style={{ backgroundColor: 'rgba(255, 255, 255, 0.3)', backdropFilter: 'blur(2px)' }}>
        <h1 className='mb-4 text-center text-5xl font-extrabold tracking-wider text-pink-900'>Memory Game</h1>

        {/* Player Name */}
        <label className='mb-2 block text-base font-bold text-white'>Player Name</label>
        <input
          value={inputName}
          onChange={handleNameChange}
          placeholder='Enter your name'
          className={`focus:outline-none mb-4 w-full rounded-lg border bg-white/80 p-2 text-gray-800 transition-colors focus:border-pink-400 focus:ring-2 focus:ring-pink-400 ${
            nameError ? 'border-red-500' : 'border-gray-300'
          }`}
        />
        {nameError && (
          <p role='alert' className='mb-2 text-xs font-medium text-red-800'>
            Name is required
          </p>
        )}

        {/* Board Size */}
        <label className='mb-2 block text-base font-bold text-white'>Board Size</label>
        <select
          value={selectedSize.name}
          onChange={(e) => {
            const size = boardSizes.find((s) => s.name === e.target.value)
            if (size) setSelectedSize(size)
          }}
          className='focus:outline-none mb-6 w-full appearance-none rounded-lg border border-gray-300 bg-white/80 p-2 text-gray-800 transition-colors focus:border-pink-400 focus:ring-2 focus:ring-pink-400'>
          {boardSizes.map((size) => (
            <option key={size.name} value={size.name}>
              {size.name}
            </option>
          ))}
        </select>

        {/* Difficulty */}
        <label className='mb-2 block text-sm font-semibold text-white'>Difficulty</label>
        <div className='relative mb-8 flex gap-3'>
          {allDifficulties.map((d) => {
            const moveLimit = getMoveLimit(d, selectedSize)
            const desc = difficultyDesc[d]
            return (
              <div key={d} className='group relative flex-1'>
                <button
                  type='button'
                  onClick={() => setDifficulty(d as Difficulty)}
                  className={`w-full rounded-lg px-3 py-2 text-sm font-semibold uppercase transition-all duration-200 ${
                    difficulty === d
                      ? 'bg-pink-700 text-white shadow-md'
                      : 'border border-gray-600 bg-gray-800 text-gray-200 hover:border-pink-500 hover:bg-gray-700 hover:text-pink-400'
                  }`}>
                  {d.charAt(0).toUpperCase() + d.slice(1)}
                </button>
                <div className='pointer-events-none absolute bottom-full left-1/2 z-10 mb-2 -translate-x-1/2 transform opacity-0 transition-opacity duration-200 group-hover:opacity-100'>
                  <div className='whitespace-nowrap rounded bg-gray-800 px-2 py-1 text-xs text-white shadow-lg'>
                    {desc} - Moves: {moveLimit}
                  </div>
                </div>
              </div>
            )
          })}
        </div>

        {/* Start Button */}
        <button
          className='w-full rounded-lg bg-pink-700 py-2 text-lg font-bold uppercase text-white shadow-lg transition-colors duration-200 hover:bg-pink-800'
          data-testid='start-button'
          onClick={handleStart}>
          Start Game
        </button>
      </div>
    </div>
  )
}

export default StartPage
