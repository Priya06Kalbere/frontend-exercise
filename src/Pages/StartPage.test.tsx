import React from 'react'
import { render, screen, fireEvent, waitFor } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import StartPage from './StartPage'

const mockNavigate = jest.fn()
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate
}))

describe('StartPage Component', () => {
  const mockOnStartGame = jest.fn()

  beforeEach(() => {
    mockOnStartGame.mockClear()
    mockNavigate.mockClear()
  })

  // Test 1: Simple smoke test with snapshot
  test('renders StartPage correctly (snapshot)', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <StartPage onStartGame={mockOnStartGame} />
      </MemoryRouter>
    )
    expect(asFragment()).toMatchSnapshot()
  })

  // Test 2: Validation check and error display
  test('shows error when starting without name (validation check)', async () => {
    render(
      <MemoryRouter>
        <StartPage onStartGame={mockOnStartGame} />
      </MemoryRouter>
    )
    const startGameButton = screen.getByTestId('start-button')
    fireEvent.click(startGameButton)
    await waitFor(() => {
      const errorMessage = screen.getByRole('alert')
      expect(errorMessage).toBeInTheDocument()
      expect(errorMessage).toHaveTextContent('Name is required')
    })
    expect(mockOnStartGame).not.toHaveBeenCalled()
  })

  // Test 3: Error removal upon typing
  test('clears error message and red border when name input is changed after error is shown', async () => {
    render(
      <MemoryRouter>
        <StartPage onStartGame={mockOnStartGame} />
      </MemoryRouter>
    )
    const nameInput = screen.getByPlaceholderText(/enter your name/i)
    fireEvent.click(screen.getByTestId('start-button'))
    await waitFor(() => {
      expect(screen.getByRole('alert')).toBeInTheDocument()
      expect(nameInput).toHaveClass('border-red-500')
    })
    fireEvent.change(nameInput, { target: { value: 'A' } })
    expect(screen.queryByRole('alert')).not.toBeInTheDocument()
    expect(nameInput).not.toHaveClass('border-red-500')
  })

  // Test 4: Successful game start with custom settings
  test('calls onStartGame with correct custom values and navigates', async () => {
    render(
      <MemoryRouter>
        <StartPage onStartGame={mockOnStartGame} />
      </MemoryRouter>
    )
    fireEvent.change(screen.getByPlaceholderText(/Enter your name/i), {
      target: { value: 'Alice' }
    })

    fireEvent.change(screen.getByRole('combobox'), {
      target: { value: '4x4' }
    })

    fireEvent.click(screen.getByRole('button', { name: /Hard/i }))
    fireEvent.click(screen.getByRole('button', { name: /Start Game/i }))
    await waitFor(() => {
      expect(mockOnStartGame).toHaveBeenCalledWith('Alice', expect.objectContaining({ rows: 4, cols: 4, name: '4x4' }), 'hard')
    })
    expect(mockNavigate).toHaveBeenCalledWith('/game')
  })

  // Test 5: Difficulty button state update
  test('updates difficulty buttons correctly upon clicking', () => {
    render(
      <MemoryRouter>
        <StartPage onStartGame={mockOnStartGame} />
      </MemoryRouter>
    )

    const easyBtn = screen.getByRole('button', { name: /Easy/i })
    const mediumBtn = screen.getByRole('button', { name: /Medium/i })
    expect(easyBtn).toHaveClass('bg-blue-500')
    expect(mediumBtn).not.toHaveClass('bg-blue-500')
    fireEvent.click(mediumBtn)
    expect(mediumBtn).toHaveClass('bg-blue-500')
    expect(easyBtn).not.toHaveClass('bg-blue-500')
  })
})
