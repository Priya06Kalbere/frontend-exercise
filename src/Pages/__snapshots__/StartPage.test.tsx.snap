// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`StartPage Component renders StartPage correctly (snapshot) 1`] = `
<DocumentFragment>
  <div
    class="h-screen flex items-center justify-center bg-gray-100"
    style="background-color: var(--bg-start);"
  >
    <div
      class="w-full max-w-md bg-white rounded-lg border shadow-lg py-10 px-8"
    >
      <h1
        class="text-2xl font-semibold text-center mb-8"
      >
        Start Game
      </h1>
      <label
        class="block text-sm mb-2"
      >
        Player Name
      </label>
      <input
        class="w-full p-2 mb-4 border rounded-md "
        placeholder="Enter your name"
        value=""
      />
      <label
        class="block text-sm mb-2"
      >
        Board Size
      </label>
      <select
        class="w-full p-2 mb-4 border rounded-md"
      >
        <option
          value="2x2"
        >
          2x2
        </option>
        <option
          value="3x4"
        >
          3x4
        </option>
        <option
          value="4x4"
        >
          4x4
        </option>
        <option
          value="4x5"
        >
          4x5
        </option>
      </select>
      <label
        class="block text-sm mb-2"
      >
        Difficulty
      </label>
      <div
        class="flex gap-2 mb-6 relative"
      >
        <div
          class="relative group flex-1"
        >
          <button
            class="w-full py-2 px-3 rounded-md border text-sm font-medium bg-blue-500 text-white border-blue-700"
            type="button"
          >
            Easy
          </button>
          <div
            class="absolute bottom-full left-1/2 transform -translate-x-1/2 mb-2 opacity-0 group-hover:opacity-100 transition-opacity duration-200 pointer-events-none z-10"
          >
            <div
              class="bg-gray-800 text-white text-xs rounded py-1 px-2 shadow-lg whitespace-nowrap"
            >
              Beginner-friendly - Moves: 36
            </div>
          </div>
        </div>
        <div
          class="relative group flex-1"
        >
          <button
            class="w-full py-2 px-3 rounded-md border text-sm font-medium bg-gray-200 border-gray-400"
            type="button"
          >
            Medium
          </button>
          <div
            class="absolute bottom-full left-1/2 transform -translate-x-1/2 mb-2 opacity-0 group-hover:opacity-100 transition-opacity duration-200 pointer-events-none z-10"
          >
            <div
              class="bg-gray-800 text-white text-xs rounded py-1 px-2 shadow-lg whitespace-nowrap"
            >
              Balanced challenge - Moves: 30
            </div>
          </div>
        </div>
        <div
          class="relative group flex-1"
        >
          <button
            class="w-full py-2 px-3 rounded-md border text-sm font-medium bg-gray-200 border-gray-400"
            type="button"
          >
            Hard
          </button>
          <div
            class="absolute bottom-full left-1/2 transform -translate-x-1/2 mb-2 opacity-0 group-hover:opacity-100 transition-opacity duration-200 pointer-events-none z-10"
          >
            <div
              class="bg-gray-800 text-white text-xs rounded py-1 px-2 shadow-lg whitespace-nowrap"
            >
              High difficulty - Moves: 24
            </div>
          </div>
        </div>
      </div>
      <button
        class="w-full py-2 bg-blue-500 text-white rounded-md hover:bg-blue-600 transition}"
        data-testid="start-button"
      >
        Start Game
      </button>
    </div>
  </div>
</DocumentFragment>
`;
